src/Analysis/Engine/Impl/DependencyResolution/PathResolverSnapshot.cs

```diff
@@ -0,0 +1,280 @@
+            lastArc = arc;
+            foreach (var name in packageNames) {
+                if (lastArc.End.IsModule) {
+                    return false;
+                }
+                var index = lastArc.End.GetChildIndex(name);
+                if (index == -1) {
+                    return false;
+                }
+                lastArc = lastArc.Append(index);
+            }
+
+            return true;
+        }
+
+        private bool TryMatchPath(in Arc lastArc, in int rootIndex, out Arc matchedLastArc) {
+            var sourceArc = lastArc.GetFirstArc();
+            matchedLastArc = new Arc(rootIndex, _roots[rootIndex]);
+
+            while (sourceArc != lastArc) {
+                sourceArc = sourceArc.Next;
+                var childIndex = matchedLastArc.End.GetChildIndex(sourceArc.End.Name);
+                if (childIndex == -1) {
+                    return false;
+                }
+                matchedLastArc = matchedLastArc.Append(childIndex);
+            }
+
+            return true;
+        }
+
+        private static Node CreateNewNodes(string modulePath, (int start, int length) unmatchedPathSpan, Node packageNode) {
+            if (packageNode != default) {
+                // Module name matches name of existing package        
+                return packageNode.ToModule(modulePath);
+            }
+
+            if (unmatchedPathSpan.start == modulePath.LastIndexOf(Path.DirectorySeparatorChar) + 1) {
+                // Module is added to existing package
+                return Node.CreateModule(modulePath.Substring(unmatchedPathSpan.start, unmatchedPathSpan.length), modulePath);
+            }
+
+            var names = modulePath.Split(Path.DirectorySeparatorChar, unmatchedPathSpan.start, unmatchedPathSpan.length);
+            var newNode = Node.CreateModule(names.Last(), modulePath);
+
+            for (var i = names.Length - 2; i >= 0; i--) {
+                newNode = new Node(names[i], newNode);
+            }
+
+            return newNode;
+        }
+
+        private static Node UpdateNodesFromEnd(Arc lastArc, Node newEnd) {
+            while (lastArc.Start != default) {
+                var newStart = lastArc.Start.ReplaceChildAt(newEnd, lastArc.EndIndex);
+                lastArc = lastArc.Previous;
+                newEnd = newStart;
+            }
+
+            return newEnd;
+        }
+
+        private bool TryFindModule(string modulePath, out Arc lastArc, out string normalizedPath, out (int start, int length) unmatchedPathSpan) {
+            if (!Path.IsPathRooted(modulePath)) {
+                throw new InvalidOperationException("Module path should be rooted");
+            }
+
+            normalizedPath = PathUtils.NormalizePath(modulePath);
+            var rootIndex = 0;
+            while (rootIndex < _roots.Length && !normalizedPath.StartsWithOrdinal(_roots[rootIndex].Name)) {
```

Comment:
OS-specific case sensitivity comparison?

Reply:
Python is case sensitive in terms of names, need to check how it works with root locations

Reply:
https://www.python.org/dev/peps/pep-0235/ specifically https://www.python.org/dev/peps/pep-0235/#id4

Reply:
PEP 235 doesn't say anything about root locations. Experiment shows that it depends on file system, so with our approximation we can use OS.